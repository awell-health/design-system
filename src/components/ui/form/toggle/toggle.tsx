import * as React from 'react';
import { cn } from '../../../../lib/utils';

interface Props extends React.InputHTMLAttributes<HTMLInputElement> {
  checked?: boolean;
  disabled?: boolean;
  label?: string;
  helpText?: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  labelClassName?: string;
  variant?: 'default' | 'positive' | 'negative';
}

const Toggle = React.forwardRef<HTMLInputElement, Props>((props, ref) => {
  const {
    className,
    checked,
    disabled,
    helpText,
    label,
    onChange,
    labelClassName,
    variant,
    ...rest
  } = props;

  return (
    <div>
      <label className='flex gap-2 cursor-pointer items-top'>
        <input
          className={cn(
            'toggle focus:bg-none',
            'border-none bg-white hover:bg-white',
            // check styles from globals.css for colors
            variant === 'positive' && 'green',
            variant === 'negative' && 'red',
            className
          )}
          checked={checked}
          ref={ref}
          {...rest}
          disabled={disabled}
          type='checkbox'
          onChange={onChange}
        />
        <span
          className={cn(
            'label-text text-slate-700 text-sm font-medium flex flex-col gap-2',
            labelClassName
          )}
        >
          {label}
          {helpText && <span className='text-slate-500'>{helpText}</span>}
        </span>
      </label>
    </div>
  );
});

Toggle.displayName = 'Toggle';

export { Toggle };
