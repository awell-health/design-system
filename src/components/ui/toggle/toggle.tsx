import * as React from 'react';
import { cva, type VariantProps } from 'class-variance-authority';

import { cn } from '@/lib/utils';

const toggleVariants = cva('toggle focus:bg-none', {
  variants: {},
  defaultVariants: {}
});

interface Props
  extends React.ButtonHTMLAttributes<HTMLInputElement>,
    VariantProps<typeof toggleVariants> {
  checked?: boolean;
  disabled?: boolean;
  label?: string;
  helpText?: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

const Toggle = React.forwardRef<HTMLInputElement, Props>(
  ({ className, checked = false, disabled = false, onChange, helpText, label, ...props }, ref) => {
    return (
      <div>
        <label className='flex gap-2 cursor-pointer'>
          <input
            className={cn(
              toggleVariants({ className }),
              // @TODO: get tailwind colors from config
              '[--tglbg:#F1F5F9] hover:[--tglbg:#E2E8F0] disabled:[--tglbg:#F1F5F9]',
              checked && '[--tglbg:#2563EB] hover:[--tglbg:#2563EB]',
              'border-none bg-white hover:bg-white'
            )}
            checked={checked}
            ref={ref}
            {...props}
            disabled={disabled}
            type='checkbox'
            onChange={onChange}
          />
          <span className='label-text text-slate-700 text-sm font-medium'>{ label }</span>
        </label>
        {helpText && (
          <div className='flex gap-2'>
            <div className='w-[48px]' />
            <span className='text-slate-500 text-sm font-normal'>{helpText}</span>
          </div>
        )}
      </div>
    );
  }
);

Toggle.displayName = 'Toggle';

export { Toggle };
