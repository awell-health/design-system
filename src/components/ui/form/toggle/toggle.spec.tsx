import { fireEvent, render } from '@testing-library/react';
import { expect, it, describe, vi } from 'vitest';
import { Toggle } from './toggle';

describe('Toggle', () => {
  const subject = (props = {}) =>
    render(<Toggle onChange={() => false} label='Toggle Label' data-testid='toggle' {...props} />);

  it('match snapshot', () => {
    const { container } = subject();

    expect(container).toMatchSnapshot();
  });

  it('match disabled snapshot', () => {
    const { container } = subject({ disabled: true });
    expect(container).toMatchSnapshot();
  });

  it('calls click handler', () => {
    const onChange = vi.fn();
    const { getByTestId } = subject({ onChange });

    fireEvent.click(getByTestId('toggle'));

    expect(onChange).toHaveBeenCalledOnce();
  });
});
