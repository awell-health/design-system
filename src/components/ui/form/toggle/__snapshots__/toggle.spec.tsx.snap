// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle > match disabled snapshot 1`] = `
<div>
  <div>
    <label
      class="flex gap-2 cursor-pointer items-top"
    >
      <input
        class="toggle focus:bg-none [--tglbg:#F1F5F9] hover:[--tglbg:#E2E8F0] disabled:[--tglbg:#F1F5F9] border-none bg-white hover:bg-white"
        data-testid="toggle"
        disabled=""
        type="checkbox"
      />
      <span
        class="label-text text-slate-700 text-sm font-medium flex flex-col gap-2"
      >
        Toggle Label
      </span>
    </label>
  </div>
</div>
`;

exports[`Toggle > match snapshot 1`] = `
<div>
  <div>
    <label
      class="flex gap-2 cursor-pointer items-top"
    >
      <input
        class="toggle focus:bg-none [--tglbg:#F1F5F9] hover:[--tglbg:#E2E8F0] disabled:[--tglbg:#F1F5F9] border-none bg-white hover:bg-white"
        data-testid="toggle"
        type="checkbox"
      />
      <span
        class="label-text text-slate-700 text-sm font-medium flex flex-col gap-2"
      >
        Toggle Label
      </span>
    </label>
  </div>
</div>
`;
